Create role based user groups.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR creates system user group based on role and add the members to the appropriate user groups along with implementing modifying the user groups when user role or `waiting_period_threshold` setting changes.
The commit for adding `group_members` field to `UserGroup` is a throwaqay commit just for implementing other parts and this will instead be implemented using a better approach to handle issues related to cycles, etc.

There are some prep commits too in tests.

Some commits that will be added -
- Migrations for creating the user groups in existing realms and adding existing users to them.
- Adding a cron job for updating the Members and Full members groups.

Posting other comments for specific doubts.

Related issue - #19525
<!-- How have you tested? -->


<!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -94,3 +94,28 @@ def get_memberships_of_users(user_group: UserGroup, members: List[UserProfile])
+    guest_system_group = role_system_groups_dict[UserProfile.ROLE_GUEST]
+
+    members_system_group.group_members.add(full_members_system_group)
+    guest_system_group.group_members.add(full_members_system_group)
```
Not sure if this approach using loop on the group lists is fine for readability or we should just add groups separetely in `group_members`

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -4735,6 +4742,16 @@ def do_change_user_role(
+    ).delete()
+
+    system_group = get_system_user_group_for_user(user_profile)
+    UserGroupMembership.objects.create(user_profile=user_profile, user_group=system_group)
```
Do we want this part in `transaction.atomic` block to avoid the groups being in inconsistent state?

timabbott replied:
You can add several `on_commit` hookos; see https://docs.djangoproject.com/en/3.2/topics/db/transactions/#order-of-execution for a brief refrence to that being supported.

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -872,6 +889,9 @@ def do_set_realm_property(
             flush_user_profile(sender=UserProfile, instance=user_profile)
 
+    if name == "waiting_period_threshold":
+        update_users_in_members_and_full_members_groups(realm)
```
Here also do we want this part within `transaction.atomic` ?

sahil839 commented:
I am not sure about the status of the adding `group_members` field, but it would be good to have a review on this so that we can atleast have an idea of implementation and potential improvements.

Also, I am not sure about where to add the daily cron job for updating members and full members user groups. I found a few, like `archieve_messages`, etc., in `/puppet/zulip/files/cron.d`, but still wanted to confirm.

I assume to implement it we can simply write a management command which will call the function that we call when changing `waiting_period_threshold` and then we can use the management command as we do for others.

sahil839 commented:
@timabbott this is ready for first review. Have left a few inline comments.

timabbott commented:
Is this ready for a next review?

sahil839 commented:
Need to change the names and add descriptions as per discussion in chat.zulip.org. Was occupied with other PRs. Will update by tommorrow.

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -128,3 +128,33 @@ def get_recursive_membership_groups(user_profile: UserProfile) -> "QuerySet[User
+
+def create_system_user_groups_for_realm(realm: Realm) -> None:
+    role_system_groups_dict: Dict[int, UserGroup] = {}
+    for role in UserGroup.SYSTEM_USER_GROUP_ROLE_MAP.keys():
```
We do not need the dict sorted actually, I mistakenly added that comment in `models.py`, removed that. The below list is used for GroupGroupMemberships and it should be sorted.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5169,6 +5173,9 @@ def do_create_realm(
         # realm, don't auto-create it to send to it; just do nothing.
         pass
+
+    create_system_user_groups_for_realm(realm)
```
I think this can and should be moved into the `transaction.atomic` block earlier in this function.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -718,6 +719,22 @@ def do_create_user(
+        )
+        do_send_user_group_members_update_event(
+            "add_members", full_members_system_group, [user_profile.id]
+        )
```
I guess ideally this would also be in a `transaction.atomic` with the user creation.  Doing so perfectly may require reworking this function to use `transaction.on_commit` for all the events it sends, etc., but we can potentially be a bit sloppier.  See `do_delete_message` for an example of this.

It'd be pretty annoying if these got out of sync, so I think it's worth at least ensuring that the `UserGroupMembership.objects.create` queries are in the main transaction.

timabbott commented on zerver/migrations/0371_create_role_based_system_groups.py:
```diff
@@ -0,0 +1,121 @@
+        group_membership_obejcts = []
+        for user in users:
+            system_group = role_system_groups_dict[user.role]
+            group_membership_obejcts.append(
```
objects has a typo in its name.

timabbott commented on zerver/migrations/0371_create_role_based_system_groups.py:
```diff
@@ -0,0 +1,121 @@
+
+    operations = [
+        migrations.RunPython(
+            create_role_based_system_groups, reverse_code=migrations.RunPython.noop, elidable=True
```
This migration seems like it might be somewhat slow, since it's O(totalusers).  @mateuszmandera I'm curious for your thoughts, but perhaps I think we can address that with:
* Setting `atomic=False`
* Doing each realm inside a transaction; that should prevent us from ending up with a realm partially created if we have a failure or ctrl+C or whatever.
* Planning to run this migration only after we've started creating these for new realms.  I think given that plan, we should make it check that the relevant user groups don't already exist in the target realm, and skip the realm if so.

mateuszmandera replied:
@timabbott @sahil839 Yeah, this plan sounds rights to me - it's typical for us to batch by realm in these kinds of migrations, with each realm being processed in its own atomic transaction

sahil839 replied:
Sorry, but it is not clear to me what will be the advantage of the third point of running the migration after we have started creating groups for new realms. Otherwise, sounds good to me.

timabbott replied:
OK, added `atomic=False`, and this block at the top of the migration:
```
+        with transaction.atomic():
+            if UserGroup.objects.filter(
+                realm=realm, name="@role:internet", is_system_group=True
+            ).exists():
+                # Handle the case where we are rerunning after a
+                # failure, and had already processed this realm.
+                continue
```

timabbott replied:
I also synced the loop to make `GroupGroupMembership` objects to use the `for supergroup in remaining_groups` loop we use in `actions.py`, and added a comment.

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -866,17 +866,18 @@ def test_register(self) -> None:
             self.register(self.nonreg_email("test"), "test")
         # Ensure the number of queries we make is not O(streams)
-        self.assert_length(queries, 89)
+        self.assert_length(queries, 95)
```
Hmm, 6 queries is more than I'd have expected; what are they?

timabbott commented on zerver/tests/test_user_groups.py:
```diff
@@ -57,8 +57,9 @@ def test_get_direct_user_groups(self) -> None:
-        self.assertEqual(user_groups[0].name, "support")
+        self.assert_length(user_groups, 3)
+        # first two user groups are role-based system groups.
+        self.assertEqual(user_groups[2].name, "support")
```
I don't like the assumption of ordering here; I think what we want instead is just to assert the set of names.

timabbott commented on zilencer/management/commands/populate_db.py:
```diff
@@ -467,6 +469,25 @@ def handle(self, *args: Any, **options: Any) -> None:
+                        user_group=group, user_profile=user_profile
+                    )
+                    user_group_memberships.append(user_group_membership)
+            UserGroupMembership.objects.bulk_create(user_group_memberships)
```
This logic is a bit messy.  It's possible we should rework how these initial users are created to go through `do_create_user` with the role passed in, but I think that's out of scope for this PR.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -833,6 +833,30 @@ def active_humans_in_realm(realm: Realm) -> Sequence[UserProfile]:
+    members_excluding_full_members = list(member_group_users - full_member_group_users)
+    new_full_members = [
+        user for user in members_excluding_full_members if not user.is_provisional_member
+    ]
```
Does this logic work consistently?  I have memories, possibly very outdated, of Django objects not behaving how one would except when placed in sets in some situations, and we generally try to work with IDs in instead for that reason.

sahil839 replied:
The test works fine. I am not sure how to check whether it works consistently or not and am not aware of any case where it could fail. Is the problem related to sets only and we can use lists instead. We would surely need objects to check `is_provisional_member` and thus ID is not enough.

sahil839 replied:
Maybe we can use `ids` for calculating `members_excluding_full_members`.

sahil839 replied:
If we want to completely avoid use of list/set of objects, we can directly use `date_joined` value in query. Code will look something like -
```
full_member_group_user_ids = full_members_system_group.direct_members.all().values_list("id", flat=True)

min_age = datetime.timedelta(days=realm.waiting_period_threshold)
new_members = full_members_system_group.direct_members.filter(
    date_joined__gt=timezone_now() - min_age
)
new_full_members = members_system_group.direct_members.filter(
    date_joined__lt=timezone_now() - min_age
).exclude(id__in=full_member_group_user_ids)

if len(new_members) > 0:
   remove_members_from_user_group(full_members_system_group, new_members)

if len(new_full_members) > 0:
   bulk_add_members_to_user_group(full_members_system_group, new_full_members)
```

timabbott replied:
OK, I've confirmed that the `set` approach will work correctly if we're careful about ensuring the sets we work with have a consistent Django model type with an appropriate primary key.

See e.g. https://github.com/django/django/commit/4668c142dce77c6f29fb75532c1acfa1b2d322ff for background.

timabbott replied:
(I'll drop the bonus commit)

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1167,6 +1168,9 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
         bulk_import_model(data, GroupGroupMembership)
 
+    if not UserGroup.objects.filter(realm=realm, is_system_group=True).exists():
+        create_and_add_users_to_system_user_groups(realm, user_profiles)
```
Let's add a comment here noting that we expect Zulip server exports to contain these, but that we have this logic here to avoid data exports from other services needing to handle system user groups.

timabbott commented:
I posted what I think is a complete batch of comments on this PR.  I think the only element with any technical risk is the transactionality stuff; feel free to consult with @mateuszmandera if you need any advice on how to do those.

(It wouldn't be a bad idea to just do a separate prep PR adjusting the existing transaction models for `do_create_user` etc., and then plan to rebase this on top of that once merged)

sahil839 commented on zerver/tests/test_signup.py:
```diff
@@ -868,17 +868,18 @@ def test_register(self) -> None:
             self.register(self.nonreg_email("test"), "test")
         # Ensure the number of queries we make is not O(streams)
-        self.assert_length(queries, 90)
+        self.assert_length(queries, 96)
```
The six queries here are - one each for getting members and full members system group, one each for creating `UserGroupMembership` objects for members and new members group and one each for `active_user_ids(user_group.realm_id)` in the `do_send_user_group_members_update_event` for adding user to member and full members groups.

sahil839 commented:
- Fixed the typo and added comments as asked above. Made comment regarding the increase in number of queries in test. Also updated test to assert set of names.
- Updated the added code to be inside a atomic transaction. Except the code in `do_set_realm_property` for which there is a PR by Mateusz to have `do_set_realm_property` in an atomic block.
- Added a temporary commit with the logic described [here](https://github.com/zulip/zulip/pull/19654#discussion_r774741529). Can keep one of the logic based on your feedback.

mateuszmandera commented on zerver/migrations/0373_create_role_based_system_groups.py:
```diff
@@ -0,0 +1,121 @@
+
+        GroupGroupMembership.objects.bulk_create(subgroup_objects)
+
+        users = UserProfile.objects.filter(realm=realm)
```
Not sure if it's a sufficiently relevant difference to make it worth, but perhaps we can just fetch `.values('id', 'role')` here since that's all that's needed rather than fetching entire UserProfiles, since this is potentially a sizeable query? Thoughts? @timabbott

timabbott replied:
Yeah, that's a good idea.

timabbott replied:
Applied this change:

```
diff --git a/zerver/migrations/0375_create_role_based_system_groups.py b/zerver/migrations/0375_create_role_based_system_groups.py
index 8a26c9231a..aee7b38f72 100644
--- a/zerver/migrations/0375_create_role_based_system_groups.py
+++ b/zerver/migrations/0375_create_role_based_system_groups.py
@@ -99,7 +99,7 @@ def create_role_based_system_groups(apps: StateApps, schema_editor: DatabaseSche
 
             GroupGroupMembership.objects.bulk_create(subgroup_objects)
 
-            users = UserProfile.objects.filter(realm=realm)
+            users = UserProfile.objects.filter(realm=realm).only("id", "role", "date_joined")
             group_membership_objects = []
             for user in users:
                 system_group = role_system_groups_dict[user.role]
@@ -118,6 +118,16 @@ def create_role_based_system_groups(apps: StateApps, schema_editor: DatabaseSche
             UserGroupMembership.objects.bulk_create(group_membership_objects)
 
 
+def delete_role_based_system_groups(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    UserGroup = apps.get_model("zerver", "UserGroup")
+    GroupGroupMembership = apps.get_model("zerver", "GroupGroupMembership")
+    UserGroupMembership = apps.get_model("zerver", "UserGroupMembership")
+    with transaction.atomic():
+        GroupGroupMembership.objects.all().delete()
+        UserGroupMembership.objects.filter(user_group__is_system_group=True).delete()
+        UserGroup.objects.filter(is_system_group=True).delete()
+
+
 class Migration(migrations.Migration):
     atomic = False
 
@@ -127,6 +137,8 @@ class Migration(migrations.Migration):
 
     operations = [
         migrations.RunPython(
-            create_role_based_system_groups, reverse_code=migrations.RunPython.noop, elidable=True
+            create_role_based_system_groups,
+            reverse_code=delete_role_based_system_groups,
+            elidable=True,
         ),
     ]
```

timabbott commented:
I've pushed an update to this PR after addressing the migration requirements.  There's a small number of things that still need attention; this should be ready to merge once those are done:
* I think we have some bugs in the life-update behavior.  Specifically, I can't convince myself that the `do_change_user_role` logic around full members doesn't incorrectly do things like deleting and then recreating an entry.  That logic also kinda duplicates `update_users_in_members_and_full_members_groups`, conceptually.
* What I'm thinking is that the right fix might be to make `update_users_in_members_and_full_members_groups` accept a `QuerySet(UserProfile)` for which users to consider, and then just call that from `do_change_user_role` for the "update the full member status of these users".  Then we'd only need to have one correct copy of this logic.
* (I guess we should add a prep commit to make `bulk_add_members_to_user_group` and its remove cousin use `transaction.atomic(savepoint=False)`?).  And `update_users_in_members_and_full_members_groups` should do the same; it'll be important for the next bullet.
* We'll also need to add a cron job to call `update_users_in_members_and_full_members_groups` in a loop over all realms on an hourly basis; otherwise users' "Full member" group membership status won't be updated properly as time passes.  This can potentially be done as a follow-up, but I don't think it's hard -- you can crib off `zerver/management/commands/soft_deactivate_users.py` to see how one adds new cron jobs.  So it might be simplest to just write it while you're handling the rest.  Probably `promote_new_full_members` would be a good name.

sahil839 commented:
Udpated the PR.
- Made `transaction.atomic` changes.
- Updated the `update_users_in_full_members_system_group` function to use it in `do_change_user_role` as well. It currently accepts `affected_user_ids` though as of now, it can only have a single user, but we can keep it as this if this is used in future. I am not sure though. Also added `ROLE_MEMBER` check to `if user.is_provisional_member or user.role != UserProfile.ROLE_MEMBER` because we need to remove user from full members group if user role is changed from a member to admin.
- Also added the cron job commit. Have added `# nocoverage` comment for the function now as am not sure how exactly we want to test it. And do we want to add logging to that command.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -847,6 +866,40 @@ def active_humans_in_realm(realm: Realm) -> Sequence[UserProfile]:
+    old_full_members = [
+        user
+        for user in full_member_group_users
+        if user.is_provisional_member or user.role != UserProfile.ROLE_MEMBER
```
Hmm, what was our decision on whether the right model is that "the full members system group directly contains all users who are at least members and not provisional" vs. "the full members system group contains all users whose role is member and are not provisional"?  

I guess they should be equivalent for all intents and purposes other than display (since the full members group contains the moderators group), and if we change our mind, the cron job will adjust within 5 minutes, so maybe it doesn't particularly matter. 

The other thought is that these queries are somewhat expensive and this cron job runs reasonably often, so I think it'd be worth refactoring the functions like `remove_members_from_user_group` to accept user IDs and have all the set work here just fetch a set of user IDs, not a set of UserProfile objects with 100 fields.

sahil839 replied:
We decided that full members group directly contains all users with role as member and are not provisional.

Also, regarding the refactoring, we also need role. Also to check `is_provisional_member` we would need to create a function which will use `role` and `date_joined` fields to find whether user is provisional or not, since we cannot access `is_provisional_member` if we instead use `values()` to get limited fields.

timabbott replied:
OK, I think this sounds complex enough it might be better as a follow-up PR, since it's just a performance optimization.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -909,6 +909,11 @@ def update_users_in_full_members_system_group(
 
+def promote_new_full_members() -> None:  # nocoverage
+    for realm in Realm.objects.all():
+        update_users_in_full_members_system_group(realm)
```
This should be `Realm.objects.filter(deactivated=False).exclude(waiting_period_threshold=None), right?  

I think we can write a great test for this function by mocking time.  But the setup is basically going to be:
(1) Hand-remove a user from the full members system group, and adding another who doesn't belong.
(2) Deactivate the realm, and verify calling this does nothing.
(3) Reactivate the realm, and verify that calling this function corrects this.

timabbott replied:
Related question: What do we want to happen with deactivated user accounts?  

The current logic updates their "Full member" status, which may be reasonable, but deserves at least a comment?  Maybe worth a bit of an API design discussion on that detail.

sahil839 replied:
> This should be `Realm.objects.filter(deactivated=False).exclude(waiting_period_threshold=0), right?

Done this change and also added the test.

> Related question: What do we want to happen with deactivated user accounts?
> 
> The current logic updates their "Full member" status, which may be reasonable, but deserves at least a comment? Maybe worth a bit of an API design discussion on that detail.

The `date_joined` field is still the same after deactivating and then reactivating the user. So, I think we should update the Full member status as well so the user is in correct group after reactivation. Can add a comment once we finally come on a conclusion.

timabbott replied:
Yeah, I think just calling the "update state" event after reactivation sounds correct.

sahil839 replied:
Just to confirm, what you mean by event here, do we want to update the user group of user only after reactivation?

timabbott commented:
I renumbered the migration and pushed back here; I expect to be able to merge it tonight or tomorrow.

sahil839 commented:
The test failure is something related to the cron job. I am not sure about how to fix it.
```
Error: /Stage[main]/Zulip::App_frontend_once/File[/etc/cron.d/promote-full-members]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///modules/zulip/cron.d/promote-full-members
```

timabbott commented:
Fixed the typo causing the production suite to fail.

timabbott commented:
Merged, huge thanks for all your work on this @sahil839!
